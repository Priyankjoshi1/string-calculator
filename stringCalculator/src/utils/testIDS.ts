export const TestIDs = {
    text: "resultString",
    submitButton: "submitButton",
    input: "inputBox"
}