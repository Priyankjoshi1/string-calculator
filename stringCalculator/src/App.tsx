import React, {useState} from 'react';
import {View, Text, TextInput, TouchableOpacity} from 'react-native';
import {TestIDs} from './utils/testIDS';
import styles from './styles/AppStyle';

const App = () => {
  const [sum, setSumValue] = useState<string>('');
  const [message, setMessage] = useState<string>('');

  const onChange = (text: string) => {
    setSumValue(text);
  };

  const getTotal = (arr: string[]) => {
    return arr.reduce((result: number, current: string): number => {
      const delimiterRegEx = /[ *_#;:/]/g; // add future delimiter in case we want to add it more
      current = current.toLowerCase().replace(delimiterRegEx, '').trim();
      current = current.toLowerCase().replace(/\n/g, ',');
      current = current.toLowerCase().replace(/,/g, ',');

      let currentVal = current.includes(',')
        ? getTotal(current?.split(','))
        : parseInt(current, 10);
      return result + currentVal;
    }, 0);
  };

  const onPress = () => {
    if (sum === undefined) {
      setMessage('You need to enter a string');
    } else if (sum === '') {
      setMessage(`The sum is 0`);
    } else {
      let stringArray = sum.split(',');
      let numArray = stringArray.map(x => parseInt(x, 10));
      let negativeNums = numArray.filter(x => x < 0);
      if (negativeNums.length > 0) {
        let negativeString = negativeNums.join(',');
        return setMessage(`negative numbers not allowed ${negativeString}`);
      }
      const result = getTotal(stringArray);
      setMessage(`The sum is ${result}`);
    }
  };

  return (
    <View style={styles.containerStyle}>
      <View style={styles.wrapperStyle}>
        <Text>Enter your numbers with separation of ","</Text>
        <TextInput
          value={sum}
          placeholder="Enter your values"
          style={styles.inputStyle}
          testID={TestIDs.input}
          onChangeText={onChange}
        />
        <TouchableOpacity
          style={styles.buttonStyle}
          onPress={onPress}
          testID={TestIDs.submitButton}>
          <Text style={styles.buttonTitleStyle}>Submit</Text>
        </TouchableOpacity>
      </View>
      <Text testID={TestIDs.text} style={styles.messageTextStyle}>
        {message}
      </Text>
    </View>
  );
};

export default App;
