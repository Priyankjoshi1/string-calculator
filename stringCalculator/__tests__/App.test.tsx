/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../src/App';
import {render, fireEvent} from '@testing-library/react-native';
import {TestIDs} from '../src/utils/testIDS';

const component = render(<App />);

describe('it to test string calculator', () => {
  it('renders correctly', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('it to have button and TextInput and result text', () => {
    const {getByTestId} = render(<App />);

    const button = getByTestId(TestIDs.submitButton);
    const input = getByTestId(TestIDs.input);
    const text = getByTestId(TestIDs.text);

    expect(button).toBeDefined();
    expect(input).toBeDefined();
    expect(text).toBeDefined();
  });

  it('give error when undefine string was there', () => {
    const {getByTestId} = render(<App />);
    const button = getByTestId(TestIDs.submitButton);
    const input = getByTestId(TestIDs.input);
    const text = getByTestId(TestIDs.text);

    fireEvent.changeText(input);
    fireEvent.press(button);
    expect(text.props.children).toBe('You need to enter a string');
  });

  it('give 0 when no string was there', () => {
    const {getByTestId} = render(<App />);
    const button = getByTestId(TestIDs.submitButton);
    const input = getByTestId(TestIDs.input);
    const text = getByTestId(TestIDs.text);

    fireEvent.changeText(input, '');
    fireEvent.press(button);
    expect(text.props.children).toBe('The sum is 0');
  });

  it('give 120 when 70,50 was there', () => {
    const {getByTestId} = render(<App />);
    const button = getByTestId(TestIDs.submitButton);
    const input = getByTestId(TestIDs.input);
    const text = getByTestId(TestIDs.text);

    fireEvent.changeText(input, '70,50');
    fireEvent.press(button);
    expect(text.props.children).toBe('The sum is 120');
  });

  it('give 1 when 1 was there', () => {
    const {getByTestId} = render(<App />);
    const button = getByTestId(TestIDs.submitButton);
    const input = getByTestId(TestIDs.input);
    const text = getByTestId(TestIDs.text);

    fireEvent.changeText(input, '1');
    fireEvent.press(button);
    expect(text.props.children).toBe('The sum is 1');
  });

  it('give 6 when 1\n2,3 was there', () => {
    const {getByTestId} = render(<App />);
    const button = getByTestId(TestIDs.submitButton);
    const input = getByTestId(TestIDs.input);
    const text = getByTestId(TestIDs.text);

    fireEvent.changeText(input, '1\n2,3');
    fireEvent.press(button);
    expect(text.props.children).toBe('The sum is 6');
  });

  it('give error when negative number was there to add', () => {
    const {getByTestId} = render(<App />);
    const button = getByTestId(TestIDs.submitButton);
    const input = getByTestId(TestIDs.input);
    const text = getByTestId(TestIDs.text);

    fireEvent.changeText(input, '-1, -2');
    fireEvent.press(button);
    expect(text.props.children).toBe('negative numbers not allowed -1,-2');

    fireEvent.changeText(input, '-1, 2');
    fireEvent.press(button);
    expect(text.props.children).toBe('negative numbers not allowed -1');
  });

  it('give 10 when given string was there', () => {
    const {getByTestId} = render(<App />);
    const button = getByTestId(TestIDs.submitButton);
    const input = getByTestId(TestIDs.input);
    const text = getByTestId(TestIDs.text);

    fireEvent.changeText(input, '\n;:\n 6,4');
    fireEvent.press(button);
    expect(text.props.children).toBe('The sum is 10');
  });

  it('give 10 when only one string was there', () => {
    const {getByTestId} = render(<App />);
    const button = getByTestId(TestIDs.submitButton);
    const input = getByTestId(TestIDs.input);
    const text = getByTestId(TestIDs.text);

    fireEvent.changeText(input, '10');
    fireEvent.press(button);
    expect(text.props.children).toBe('The sum is 10');
  });
});
