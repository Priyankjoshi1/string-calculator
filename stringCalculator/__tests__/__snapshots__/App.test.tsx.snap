// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it to test string calculator renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "flex-start",
      "marginTop": 100,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Text>
      Enter your numbers with separation of ","
    </Text>
    <TextInput
      onChangeText={[Function]}
      placeholder="Enter your values"
      style={
        {
          "backgroundColor": "#ebebeb",
          "borderRadius": 20,
          "height": 50,
          "marginHorizontal": 30,
          "marginTop": 20,
          "textAlign": "center",
          "width": "80%",
        }
      }
      testID="inputBox"
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#0000ff",
          "borderRadius": 20,
          "height": 50,
          "justifyContent": "center",
          "marginTop": 50,
          "opacity": 1,
          "width": "80%",
        }
      }
      testID="submitButton"
    >
      <Text
        style={
          {
            "color": "#ffffff",
          }
        }
      >
        Submit
      </Text>
    </View>
  </View>
  <Text
    style={
      {
        "marginTop": 100,
      }
    }
    testID="resultString"
  />
</View>
`;
